src/com/xilinx/rapidwright/rwroute/RWRoute.java

```diff
@@ -0,0 +1,1843 @@
+		Set<NetWrapper> illegalRoutes = this.findIllegalRoutes();
+		// fix routes with cycles and / or multi-driver nodes
+		for(NetWrapper route:illegalRoutes){
+			for(Connection c:route.getConnection()){
+				try{
+					if(!c.isDirect()) this.ripUp(c);
+				}catch (Exception e){
+					e.printStackTrace();
+				}
+			}
+			RouteFixer graphHelper = new RouteFixer(route, this.rnodesCreated);
+			graphHelper.finalizeRoutesOfConnections();
+		}
+		return illegalRoutes;
+	}
+	
+	/**
+	 * Finds nets that have illegal routes by checking its connections' routes.
+	 * @return A set of routed {@link NetWrapper} instances whose should be fixed.
+	 */
+	private Set<NetWrapper> findIllegalRoutes(){
+		Set<NetWrapper> illegalRoutes = new HashSet<>();
+		for(NetWrapper net : this.nets) {
+			this.buildDriverCountsOfRnodes(net);
+			for(Connection con : net.getConnection()) {
+				if(this.shouldMergePath(con)) {
+					illegalRoutes.add(net);
+					if(config.isTimingDriven()) this.addNodesDelays(net);
+					break;
+				}
+			}
+		}
+		return illegalRoutes;
+	}
+	
+	/**
+	 * Builds the driversCounts map of each {@link Routable} instance that is used by a net.
+	 * @param netWrapper A NetWrapper instance that represents a net.
+	 */
+	private void buildDriverCountsOfRnodes(NetWrapper netWrapper) {
+		for(Connection connection : netWrapper.getConnection()) {
+			Routable driver = null;
+			for(int i = connection.getRnodes().size() - 1; i >= 0; i--){
+				Routable rnode = connection.getRnodes().get(i);
+				if(driver == null){
+					driver = rnode;
+				}else{
+					rnode.addDriver(driver);
+					driver = rnode;
+				}
+			}
+		}
+	}
+	
+	/**
+	 * Adds nodes and delay values of a routed to the map.
+	 * @param net The routed net.
+	 */
+	private void addNodesDelays(NetWrapper net){	
+		for(Connection c:net.getConnection()){
+			for(Routable group : c.getRnodes()){
+				this.nodesDelays.put(group.getNode(), group.getDelay());
+			}
+		}
+	}
+	
+	/**
+	 * Checks if a connection has multi-driver nodes.
+	 * @param connection The connection in question.
+	 * @return true, if the connection has multi-driver nodes.
+	 */
+	private boolean shouldMergePath(Connection connection) {
+		return connection.useRnodesWithMultiDrivers();
+	}
+	
+	/**
+	 * Rips up a connection.
+	 * @param connection The connection to be ripped up.
+	 */
+	private void ripUp(Connection connection){
+		for(int i = connection.getRnodes().size() - 1; i >= 0; i--){
+			Routable rnode = connection.getRnodes().get(i);
+			rnode.reduceConnectionCountOfUser(connection.getSource());
+			rnode.reduceConnectionCountOfUser(connection.getNetWrapper().getOldSource());
+			rnode.updatePresentCongesCost(this.presentCongesFac);
+		}
+	}
+	
+	/**
+	 * Updates the users and present congestion cost of rnodes used by a routed connection.
+	 * @param connection The routed connection.
+	 */
+	private void updateUsersAndPresentCongesCost(Connection connection){
+		for(int i = connection.getRnodes().size()-1; i >= 0; i--){
```

Comment:
Is it important to iterate backwards here?

Reply:
Not really. Updated in the new commit.